WEB3-644: WEB3-560: feat: Helpers for queries over multiple blocks of the same chain

Wollac opened the pull request.

Wollac committed:
add multiblock env

Wollac committed:
fix doc

Wollac committed:
explain reason for allow(private_bounds)

Wollac committed:
update CHANGELOG.md

Wollac committed:
minor improvements

Wollac committed:
use MultiblockEvmEnv

Wollac committed:
shorten log

Wollac committed:
fix clippy warning

Wollac committed:
export history consts

Wollac committed:
Merge branch 'main' into feat/multiblock

Wollac committed:
add tests for all commitments

Wollac committed:
add len

nategraf reviewed (commented):
I think this looks pretty good, and a definite improvement over the version I put together. I think this will be quite nice for consumers, and should avoid them needing to use `SteelVerifier` directly. The `token-stats` example looks quite nice with this.

Wollac committed:
add is_empty

Wollac committed:
Update crates/steel/src/multiblock.rs

Co-authored-by: Victor Snyder-Graf <victor@risczero.com>

Wollac committed:
nit

Wollac commented on crates/steel/src/host/builder.rs:
```diff
@@ -250,6 +270,22 @@ impl<P, F, S, B> EvmEnvBuilder<P, F, S, B> {
         self
     }
 
+    /// Returns a copy of the builder with elided commitment config and set EVM execution block.
+    pub(crate) fn to_block(&self, block: impl Into<BlockId>) -> EvmEnvBuilder<P, F, S, ()>
+    where
+        P: Clone,
+        S: Clone,
+    {
+        EvmEnvBuilder {
+            provider: self.provider.clone(),
+            provider_config: self.provider_config.clone(),
+            block: block.into(),
+            chain_spec: self.chain_spec.clone(),
+            commitment_config: (),
+            phantom: PhantomData,
+        }
+    }
```
According to the Rust naming conventions `to_` is borrowed -> owned for non-Copy types, like [`Path::to_str`](https://doc.rust-lang.org/std/path/struct.Path.html#method.to_str) or [`str::to_lowercase()`](https://doc.rust-lang.org/std/primitive.str.html#method.to_lowercase).
So the alternative would be to rename this to an `into_` and combine it will an explicit `clone`, I'll check whether that might help readability, even though it will always be combined with a clone.

Wollac commented on crates/steel/src/multiblock.rs:
```diff
@@ -0,0 +1,430 @@
+                &self,
+                env: HostEvmEnv<ProviderDb<N, P>, F, ()>,
+            ) -> anyhow::Result<EvmInput<F>> {
+                // ignore the builder and use the available env with block commitment directly
+                env.into_input().await
```
> It seems your intention is essentially to take the EVM data (e.g. storage and event inclusion proofs) and combine it with the commitment provided by self.

Exactly. Use the builder `self` to attach the correct commitment to the evm data, and also convert it into a type `EvmImput`, that does not depend on an generics.

> However, both in this case and the other implementations, self and env both contain the EVM data and the commitment.

No, `self` is a builder, so by definition it does not contain any EVM data and just config and  `env` is a `HostEvmEnv<D, F, ()>`, while the final `()` actually assures that it does not contain any specific commitment.

I agree that the entire `InputBuilder` design is not ideal. However, it attaches different commitments to EVM data and tries to minimize code duplication, which is exactly what it is supposed to do.

Wollac commented on crates/steel/src/multiblock.rs:
```diff
@@ -0,0 +1,430 @@
+        <F::Receipt as TryFrom<<N as Network>::ReceiptResponse>>::Error: Display,
+        EvmEnvBuilder<P, F, &'a ChainSpec<F::SpecId>, B>: InputBuilder<ProviderDb<N, P>, F>,
+    {
+        /// Creates a new [HostMultiblockEvmEnv] from the given [EvmEnvBuilder].
+        ///
+        /// This ignores any potential EVM execution block set in the builder, but all other options
+        /// specified with this builder are incorporated when creating the individual [EvmEnv].
+        pub fn from_builder(builder: EvmEnvBuilder<P, F, &'a ChainSpec<F::SpecId>, B>) -> Self {
+            Self {
+                builder,
+                env: MultiblockEvmEnv(BTreeMap::new()),
+            }
+        }
```
Honestly, I like this design because we get the staged builder pattern from the builder for free. Of course, we can make HostMultiblockEvmEnv the builder, but then we'd have to still double all these different config steps as well as the different commitment types. Ii would probably be worth it if this would allows us to get rid of the InputBuilder trait but I dont see that.

Wollac replied:
In order to make the HostMultiblockEvmEnv a sufficient builder, we would need to add all the following
```rust
    provider: P,
    provider_config: ProviderConfig,
    block: BlockId,
    chain_spec: S,
    commitment_config: B,
    phantom: PhantomData<F>,
```

and also implement staged builder pattern for this. This would lead to big code duplication. 
I also think that the pattern
```rust
let builder = EthEvmEnv::builder().provider(p).chain_spec(&ETH_MAINNET_CHAIN_SPEC);
let mut envs = HostMultiblockEvmEnv::from_builder(builder);
```
is closer to our existing pattern of how to build a HostEvmEnv.  Thus, I have kept this design.

Wollac committed:
improve token-stats example

Wollac committed:
cleanups

Wollac committed:
fix license header

Wollac committed:
add block number

Wollac commented on examples/token-stats/methods/guest/src/main.rs:
```diff
@@ -24,36 +24,25 @@ const SECONDS_PER_YEAR: u64 = 60 * 60 * 24 * 365;
-        .call();
+    let envs = input.into_env(&ETH_MAINNET_CHAIN_SPEC);
+
+    // Execute the view calls on each EVM state.
```
I have changed the example so that the environments must be exactly 7,200 blocks (24 hours) apart, and it commits to the number of such intervals.

Wollac committed:
simplify input builder

Wollac committed:
add STEEL_TEST_CANCUN_CHAIN_SPEC

Wollac committed:
add STEEL_TEST_CANCUN_CHAIN_SPEC

Wollac committed:
cargo fmt

Wollac committed:
sort test chain configs

Wollac committed:
Merge branch 'main' into feat/multiblock

Wollac committed:
fix unused imports

Wollac committed:
Merge branch 'main' into feat/multiblock

Wollac committed:
add eth aliases

Wollac committed:
revert Steel.sol changes

Wollac committed:
fix clippy warnings

Wollac committed:
Revert "fix clippy warnings"

This reverts commit d0131214e0ac8284d3ae56e6bcf3a217370b5834.

Wollac committed:
cleanups

Wollac committed:
Merge branch 'main' into feat/multiblock

Wollac committed:
minor cleanups

Wollac committed:
add build_multi

Wollac committed:
minor cleanups

Wollac committed:
fix stale comment

Wollac committed:
update changelog

Wollac committed:
update license header

Wollac committed:
use build_multi

Wollac committed:
update license header

Wollac committed:
Merge remote-tracking branch 'origin/main' into feat/multiblock

# Conflicts:
#	CHANGELOG.md

Wollac committed:
minor cleanups

Wollac committed:
fix: rustfmt in token-stats guest

Wollac committed:
Merge remote-tracking branch 'origin/main' into feat/multiblock

# Conflicts:
#	crates/steel/src/lib.rs

Wollac committed:
chore: address review feedback on multiblock env

- warn instead of silently building an input with no recorded state
  access; the explicit guard was removed because multiblock legitimately
  builds commitment-only envs and the parent-hash fast path records none
- document why the EIP-2935 spacing check inspects current_env (storage
  is not backfilled at activation, so a pre-fork hash is unretrievable)
- drop the dead .context() on the infallible build_input! merge so both
  merge sites use a bare ?
- token-stats: fix stale README output, use .expect for the guest rate
  conversion, and consolidate the 7200 block interval into a shared
  BLOCK_INTERVAL const

Wollac committed:
chore: bump copyright year on files modified in 2026

Wollac committed:
fix(examples): switch OP examples from Base Mainnet to OP Mainnet

Alchemy currently fails all eth_getProof requests on base-mainnet,
breaking the preflight of the op examples in CI. OP Mainnet (as well as
both Sepolia networks) serve proofs fine, so move the l2 and l2-to-l1
examples to OP Mainnet. This keeps L1 on Ethereum Mainnet, so the
shared L1_RPC_URL and the l1-to-l2 example are unaffected.

Wollac committed:
fix: address multiblock review feedback

- reject the zero value returned by the EIP-2935 history storage and
  EIP-4788 beacon roots contracts for ring buffer slots that were never
  written; block commitments with an unset slot fall back to BLOCKHASH
  verification within the 256-block window
- tighten EvmEnv::merge back to a single commitment type; the cross-type
  state merge needed by multiblock is now pub(crate) merge_state
- reject blocks at or after an explicit commitment target in get_or_build
  before any RPC, and warn when build_multi ignores a configured block
- document the beacon chain-head and commitment-target constraints; clarify
  the beacon "no child block" error for the multiblock case
- downgrade the forgotten-preflight host error to a warning and make the
  guest MPT panic self-diagnosing; restore the no_preflight test
- unify commitment version rendering via Commitment::version_name
- token-stats: require >= 2 samples and rename the journal field to numDays
  to avoid the Solidity reserved keyword

Wollac committed:
style: apply default rustfmt formatting

Wollac committed:
style: apply default rustfmt to token-stats guest

Wollac committed:
chore: bump copyright year to 2026 on files modified in 2026
